File: src/com/suushiemaniac/cubing/alglib/transform/rotate/CubicRotation.java
package com.suushiemaniac.cubing.alglib.transform.rotate;

import com.suushiemaniac.cubing.alglib.move.CubicMove;
import com.suushiemaniac.cubing.alglib.move.Move;
import com.suushiemaniac.cubing.alglib.move.modifier.CubicModifier;
import com.suushiemaniac.cubing.alglib.move.plane.CubicPlane;
import com.suushiemaniac.cubing.alglib.util.ArrayUtils;

import static com.suushiemaniac.cubing.alglib.move.plane.CubicPlane.*;

public enum CubicRotation implements Rotation {
    Y_90, X_90, Z_90, Y_180, X_180, Z_180, Y_270, X_270, Z_270;

    private static CubicPlane[][][] rotOrder = {
            {{RIGHT, FRONT, LEFT, BACK}, {RIGHT_SLICE, FRONT_SLICE, LEFT_SLICE, BACK_SLICE}, {MIDDLE, SANDWICH}, {SPATIAL, HORIZONTAL}},
            {{UP, BACK, DOWN, FRONT}, {UP_SLICE, BACK_SLICE, DOWN_SLICE, FRONT_SLICE}, {SANDWICH, EQUATOR}, {VERTICAL, SPATIAL}},
            {{RIGHT, DOWN, LEFT, UP}, {RIGHT_SLICE, DOWN_SLICE, LEFT_SLICE, UP_SLICE}, {MIDDLE, EQUATOR}, {HORIZONTAL, VERTICAL}}
    };

    @Override
    public Move rotate(Move origin) {
        if (!(origin instanceof CubicMove)) return null;
        else {
            int rotBaseOrdinal = this.ordinal() % 3;
            int rotAmountOrdinal = this.ordinal() / 3;

            CubicMove cOrigin = (CubicMove) origin;

            CubicPlane cPlane = cOrigin.getPlane();
            int ordinal = cPlane.ordinal();
            int base = ordinal < 6 ? 0 : ordinal < 12 ? 6 : ordinal < 15 ? 12 : 15;
            int size = base < 12 ? 6 : 3;
            boolean isPlaneAffected = rotBaseOrdinal != (ordinal - base) / (size / 3);

            int typeBase = ordinal < 6 ? 0 : ordinal < 12 ? 1 : ordinal < 15 ? 2 : 3;
            int rotIndex = isPlaneAffected ? ArrayUtils.arrayIndex(rotOrder[rotBaseOrdinal][typeBase], cPlane) : 0;
            int modLen = (size / 3) * 2;
            int fullIndex = (rotIndex + rotAmountOrdinal + 1) % modLen;
            CubicPlane plane = isPlaneAffected ? rotOrder[rotBaseOrdinal][typeBase][fullIndex] : cPlane;

            CubicModifier modifier = cOrigin.getModifier();
            for (int i = 0; i <= rotAmountOrdinal; i++)
                if (isPlaneAffected && modLen == 2 && (++rotIndex % 2) % 2 == 1)
                    modifier = modifier.inverse();

            return new CubicMove(plane, modifier, cOrigin.getDepth());
        }
    }

    @Override
    public Move transform(Move origin) {
        return this.rotate(origin);
    }
}

File: src/main/kotlin/com/suushiemaniac/cubing/alglib/transform/rotate/CubicRotation.kt
package com.suushiemaniac.cubing.alglib.transform.rotate

import com.suushiemaniac.cubing.alglib.move.CubicMove
import com.suushiemaniac.cubing.alglib.move.Move
import com.suushiemaniac.cubing.alglib.move.plane.CubicPlane.*

enum class CubicRotation : Rotation {
    Y_90, X_90, Z_90, Y_180, X_180, Z_180, Y_270, X_270, Z_270;

    override fun rotate(origin: Move): Move {
        if (origin !is CubicMove)
            return origin
        else {
            val rotBaseOrdinal = this.ordinal % 3
            val rotAmountOrdinal = this.ordinal / 3

            val cPlane = origin.plane
            val ordinal = cPlane.ordinal
            val base = if (ordinal < 6) 0 else if (ordinal < 12) 6 else if (ordinal < 15) 12 else 15
            val size = if (base < 12) 6 else 3
            val isPlaneAffected = rotBaseOrdinal != (ordinal - base) / (size / 3)

            val typeBase = if (ordinal < 6) 0 else if (ordinal < 12) 1 else if (ordinal < 15) 2 else 3
            var rotIndex = if (isPlaneAffected) ROT_ORDER[rotBaseOrdinal][typeBase].indexOf(cPlane) else 0
            val modLen = size / 3 * 2
            val fullIndex = (rotIndex + rotAmountOrdinal + 1) % modLen
            val plane = if (isPlaneAffected) ROT_ORDER[rotBaseOrdinal][typeBase][fullIndex] else cPlane

            var modifier = origin.modifier
            for (i in 0..rotAmountOrdinal)
                if (isPlaneAffected && modLen == 2 && ++rotIndex % 2 % 2 == 1)
                    modifier = modifier.inverse()

            return CubicMove(plane, modifier, origin.depth, origin.fromDepth)
        }
    }

    companion object {
        private val ROT_ORDER = arrayOf(
                arrayOf(
                        arrayOf(RIGHT, FRONT, LEFT, BACK),
                        arrayOf(MIDDLE, SANDWICH),
                        arrayOf(SPATIAL, HORIZONTAL)
                ),
                arrayOf(
                        arrayOf(UP, BACK, DOWN, FRONT),
                        arrayOf(SANDWICH, EQUATOR),
                        arrayOf(VERTICAL, SPATIAL)
                ),
                arrayOf(
                        arrayOf(RIGHT, DOWN, LEFT, UP),
                        arrayOf(MIDDLE, EQUATOR),
                        arrayOf(HORIZONTAL, VERTICAL)
                )
        )
    }
}
